Add PUT traces endpoint and functionality match_and_apply

gretelliz opened the pull request:
This PR is related to issues #43 #44 #45 #46
It is now possible to request stored flows from flow_manager.
`match_an_apply` has been moved from flow_stats.
Added new `PUT /traces` endpoint for bulk requests.

This endpoint returns a list of traces organized by switches. The specification is added in openapi.yml.
Examples of requests and responses: 

```
[
  {
    "trace": {
      "switch": {
        "dpid": "00:00:00:00:00:00:00:01",
        "in_port": 1
      },
      "eth": {
        "dl_vlan": 100,
      },
    }
  },
  {
    "trace": {
      "switch": {
        "dpid": "00:00:00:00:00:00:00:03",
        "in_port": 1
      },
      "eth": {
        "dl_vlan": 100,
       },
    }
  }
]
```

Response:

```
{
    "00:00:00:00:00:00:00:01": [
        [
            {
                "dpid": "00:00:00:00:00:00:00:01",
                "port": 1,
                "time": "2022-10-13 13:19:36.684592",
                "type": "starting",
                "vlan": 100
            },
            {
                "dpid": "00:00:00:00:00:00:00:02",
                "port": 2,
                "time": "2022-10-13 13:19:36.684857",
                "type": "trace",
                "vlan": 2813
            },
            {
                "dpid": "00:00:00:00:00:00:00:03",
                "port": 2,
                "time": "2022-10-13 13:19:36.685209",
                "type": "trace",
                "vlan": 426
            }
        ]
    ],
    "00:00:00:00:00:00:00:03": [
        [
            {
                "dpid": "00:00:00:00:00:00:00:03",
                "port": 1,
                "time": "2022-10-13 13:19:36.685563",
                "type": "starting",
                "vlan": 100
            },
            {
                "dpid": "00:00:00:00:00:00:00:02",
                "port": 3,
                "time": "2022-10-13 13:19:36.685718",
                "type": "trace",
                "vlan": 4001
            },
            {
                "dpid": "00:00:00:00:00:00:00:01",
                "port": 2,
                "time": "2022-10-13 13:19:36.685955",
                "type": "trace",
                "vlan": 3824
            }
        ]
    ]
}
```

It has also been validated that one response is only returned for equal elements within a list of requests.
The previous case, where only one trace is requested, remains with the same structure.

Please consider function [map_flows](https://github.com/kytos-ng/sdntrace_cp/blob/3de412829e95479f921bf9397b1ec41d28963795/main.py#L166), and comment if you agree.

viniarck reviewed (changes requested):
@gretelliz great to have new new `PUT /traces` endpoint, we'll need to adapt certain implementations detail though, left some threads open to sync with you. 

Please, update also your PR to have a meaningful title to facilitate search and easier for reviewers to identify it, no need to change the branch name but in a next PR make sure to use like `feature/<some_feature>` `fix/<some_fix>` or any other meaningful names or patterns, we don't have much guidelines for that yet, but that helps, thanks. Maybe this is something that we could also improve in terms of contribution docs for newcomers https://github.com/kytos-ng/kytos-ng.github.io/issues/24.

gretelliz commented:
Hi @viniarck I have done changes according to your suggestions.

There is now a separate new function for the `/traces` endpoint.
I removed the variable self.stored_flows.
I also removed the translation to vlan_vid.
I removed the lots of nested of ifs as you suggested.

viniarck commented on main.py:
```diff
@@ -58,31 +61,48 @@ def shutdown(self):
     def trace(self):
         """Trace a path."""
         entries = request.get_json()
+        entries = convert_entries(entries)
```
Gretel, let's get rid of any of the translations https://github.com/kytos-ng/sdntrace_cp/pull/47#discussion_r995759094, we don't need them, keeping them since we'll nuke `GenericFlow` will only make the code more difficult to maintain.

viniarck commented on main.py:
```diff
@@ -140,3 +160,102 @@ def update_circuits(self, event):
+                first flow or not
+        :return: If many, the list of matched flows, or the matched flow
+        """
+        stored_flows = self.get_stored_flows()
```
`match_flows` is called iterativelly, and every time you're refetching all stored_flows from all dpids, so to be aligned with the strategy of bulk request, let's pre-fetch all the stored flows and pass them and then you iterate on them. We need this endpoint to perform well and be optimized since mef_eline will rely on it to run sdntrace for hundreds of EVCs periodically.

gretelliz replied:
Fine @viniarck, that's why before I used the self.store_flows variable that I updated every time the endpoint was called. Now I have done it with a variable that is passed by parameter.

gretelliz replied:
Vinicius, thanks, you are right.

viniarck reviewed (commented):
@gretelliz https://github.com/kytos-ng/sdntrace_cp/pull/47#issuecomment-1282539181, 

Thanks for pushing the last commits, nice to see how this is shaping up, I've just reviewed and followed up. Also, when it's ready for another review please explore it locally with `mef_eline` with an EVC with both `PUT /traces` and `PUT /trace` endpoints and let us know how it went.

viniarck commented:
> @gretelliz [#47 (comment)](https://github.com/kytos-ng/sdntrace_cp/pull/47#issuecomment-1282539181),
> 
> Thanks for pushing the last commits, nice to see how this is shaping up, I've just reviewed and followed up. Also, when it's ready for another review please explore it locally with `mef_eline` with an EVC with both `PUT /traces` and `PUT /trace` endpoints and let us know how it went.

@gretelliz thanks for pushing these latest commits, I've just closed the threads you've addressed. The only thing renaming is to explore locally and ideally re-run existing e2e tests related to sdntrace_cp, it'd be great to also have e2e coverage for this new endpoint.

gretelliz commented:
I explored the last commit locally with an EVC for both PUT /traces and PUT /trace endpoints:

It seems to be fine. Here you can see some examples:

Steps:

1. Create a new circuit:

request:
```
{
            "name": "Vlan_400" ,
            "enabled": true,
            "dynamic_backup_path": true,
            "uni_a": {
                "interface_id": "00:00:00:00:00:00:00:01:1"
            },
            "uni_z": {
                "interface_id": "00:00:00:00:00:00:00:03:1"
            }
}
```

response:
```
{
    "circuit_id": "d05f0d7128c447"
}
```

2. Test `PUT /trace` endpoint:

request:
```
  {
    "trace": {
      "switch": {
        "dpid": "00:00:00:00:00:00:00:01",
        "in_port": 1
      },
      "eth": {
        "dl_vlan": 100
      }
    }
  }
```

response:
```
{
    "result": [
        {
            "dpid": "00:00:00:00:00:00:00:01",
            "port": 1,
            "time": "2022-10-19 14:11:44.163363",
            "type": "starting",
            "vlan": 100
        },
        {
            "dpid": "00:00:00:00:00:00:00:02",
            "port": 2,
            "time": "2022-10-19 14:11:44.163428",
            "type": "trace",
            "vlan": 1762
        }
    ]
}
```

3. Test `PUT /traces` endpoint:

request 1: Different entries
```
[
  {
    "trace": {
      "switch": {
        "dpid": "00:00:00:00:00:00:00:01",
        "in_port": 1
      },
      "eth": {
        "dl_vlan": 100
      }
    }
  },
  {
    "trace": {
      "switch": {
        "dpid": "00:00:00:00:00:00:00:03",
        "in_port": 1
      },
      "eth": {
        "dl_vlan": 100
      }
    }
  }
]
```

response: 
```
{
    "00:00:00:00:00:00:00:01": [
        [
            {
                "dpid": "00:00:00:00:00:00:00:01",
                "port": 1,
                "time": "2022-10-19 14:14:42.673497",
                "type": "starting",
                "vlan": 100
            },
            {
                "dpid": "00:00:00:00:00:00:00:02",
                "port": 2,
                "time": "2022-10-19 14:14:42.673548",
                "type": "trace",
                "vlan": 1762
            }
        ]
    ],
    "00:00:00:00:00:00:00:03": [
        [
            {
                "dpid": "00:00:00:00:00:00:00:03",
                "port": 1,
                "time": "2022-10-19 14:14:42.673569",
                "type": "starting",
                "vlan": 100
            },
            {
                "dpid": "00:00:00:00:00:00:00:02",
                "port": 3,
                "time": "2022-10-19 14:14:42.673583",
                "type": "trace",
                "vlan": 279
            },
            {
                "dpid": "00:00:00:00:00:00:00:01",
                "port": 2,
                "time": "2022-10-19 14:14:42.673605",
                "type": "trace",
                "vlan": 1458
            }
        ]
    ]
}
```

request 2: Same dpid and different ports in entries
```
[
  {
    "trace": {
      "switch": {
        "dpid": "00:00:00:00:00:00:00:01",
        "in_port": 2
      },
      "eth": {
        "dl_vlan": 100
      }
    }
  },
  {
    "trace": {
      "switch": {
        "dpid": "00:00:00:00:00:00:00:01",
        "in_port": 1
      },
      "eth": {
        "dl_vlan": 100
      }
    }
  }
]
```

response:
```
{
    "00:00:00:00:00:00:00:01": [
        [
            {
                "dpid": "00:00:00:00:00:00:00:01",
                "port": 2,
                "time": "2022-10-19 14:16:58.703696",
                "type": "starting",
                "vlan": 100
            }
        ],
        [
            {
                "dpid": "00:00:00:00:00:00:00:01",
                "port": 1,
                "time": "2022-10-19 14:16:58.703841",
                "type": "starting",
                "vlan": 100
            },
            {
                "dpid": "00:00:00:00:00:00:00:02",
                "port": 2,
                "time": "2022-10-19 14:16:58.703872",
                "type": "trace",
                "vlan": 1762
            }
        ]
    ]
}
```

request 3: Same entry twice
```
[
  {
    "trace": {
      "switch": {
        "dpid": "00:00:00:00:00:00:00:01",
        "in_port": 1
      },
      "eth": {
        "dl_vlan": 100
      }
    }
  },
  {
    "trace": {
      "switch": {
        "dpid": "00:00:00:00:00:00:00:01",
        "in_port": 1
      },
      "eth": {
        "dl_vlan": 100
      }
    }
  }
]
```

response:
```
{
    "00:00:00:00:00:00:00:01": [
        [
            {
                "dpid": "00:00:00:00:00:00:00:01",
                "port": 1,
                "time": "2022-10-19 14:17:41.065475",
                "type": "starting",
                "vlan": 100
            },
            {
                "dpid": "00:00:00:00:00:00:00:02",
                "port": 2,
                "time": "2022-10-19 14:17:41.065531",
                "type": "trace",
                "vlan": 1762
            }
        ]
    ]
}
```

viniarck reviewed (approved):
Nicely done @gretelliz. I'll leave the PR approved, it's up to you if you want to also ship an e2e test now (with this PR getting merged soon) or later on. Also, [e2e sdntrace_cp test case](https://github.com/amlight/kytos-end-to-end-tests/blob/master/tests/test_e2e_40_sdntrace.py#L240-L241) that used to be marked with `xfail` shouldn't be failing anymore since it's computing the trace from flow_manager stored_flows.

gretelliz commented:
Great! @viniarck. I will close this PR here and open a new issue for the e2e test.
